### PR TITLE
[WIP] support different tokenizers

### PR DESCRIPTION
**Pull request by H-Huang**
Requesting feedback since this will change the tokenizer directory for existing users. We need to add support more multiple tokenizers in titan (we currently only have the one used in llama3)

Current workflow:
- Users call `scripts/download_tokenizer.py`
- Saves the tokenizer to `assets/tokenizer/original/tokenizer.model`
- Users uses the tokenizer by adding `tokenizer.model` in the `.toml` configs (https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/train_configs/llama3_8b.toml#L21
- We retrieve `tokenizer.model` and wrap with our own `tiktoken` tokenizer (https://github.com/pytorch/torchtitan/blob/main/torchtitan/datasets/tokenizer/tiktoken.py)

New workflow:
- Users call `scripts/download_tokenizer.py`
- Saves tokenizer configs to a directory named `assets/tokenizer/<model_name>/`
- Users uses the tokenizer by referencing the directory mentioned in the previous step in the `.toml` configs 
- Using hugging face tokenizer lib to load the tokenizer

Pros:
- Supports additional/existing preloaded tokenizers
- Can remove tiktoken dependency

Cons:
- Breaks current users depending on `original/tokenizer.model`
- Adds new dependency on HF tokenizer

### CODE REVIEW

File: scripts/use_tokenizer_example.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
+#!/usr/bin/env python3
```
Comment (H-Huang) [inline scripts/use_tokenizer_example.py]: Will remove this file, i was just using to test

File: scripts/download_tokenizer.py [hunk 2]
```diff
@@ -4,28 +4,75 @@
...
+        hf_token: Optional HuggingFace API token for private repos
+    """
+    from huggingface_hub import hf_hub_download, list_repo_files
+    import os
+    
+    # Extract model name from repo_id (part after "/")
+    model_name = repo_id.split("/")[-1]
+    model_dir = os.path.join(local_dir, model_name)
+    
+    # Common tokenizer files to download
```
Comment (wwwjn) [inline scripts/download_tokenizer.py]: This might not be comprehensive and need user's double-check, eg, CLIP tokenizer used in FLUX needs `special_tokens_map.json` and `vocab.json`: https://huggingface.co/openai/clip-vit-large-patch14/tree/main

Comment (H-Huang) [reply scripts/download_tokenizer.py]: Thanks, good point! Let me look into how image tokenizers work too

File: README.md [hunk 3]
```diff
@@ -103,8 +103,11 @@ Once you have confirmed access, you can run the following command to download th
 ```bash
 # Get your HF token from https://huggingface.co/settings/tokens
 
-# Llama 3.1 tokenizer.model
-python scripts/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3.1-8B --tokenizer_path "original" --hf_token=...
+# Llama 3.1 tokenizer (automatically downloads original/tokenizer.model)
+python scripts/download_tokenizer.py --repo_id meta-llama/Meta-Llama-3.1-8B --hf_token=...
+
+# DeepSeek tokenizer (automatically downloads tokenizer.json and tokenizer_config.json)
+python scripts/download_tokenizer.py --repo_id deepseek-ai/DeepSeek-V3
```
Comment (fegin) [inline README.md]: Just curious, is `hf_token` not required?

File: scripts/download_tokenizer.py (hunk 2 above)
Comment (wwwjn) [reply scripts/download_tokenizer.py]: CLIP is a text encoder instead of image encoder, and I feel like the "transformers" repo supports more tokenizers than the `tokenizer` repo

File: README.md (hunk 3 above)
Comment (H-Huang) [reply README.md]: yup good Q, it is MIT license so no token required: https://huggingface.co/deepseek-ai/DeepSeek-V3/blob/main/tokenizer.json

File: scripts/download_tokenizer.py (hunk 2 above)
Comment (H-Huang) [reply scripts/download_tokenizer.py]: I was looking into why there are two version of "tokenizer", it turns out the standalone tokenizer library is that "fast" version (https://huggingface.co/docs/transformers/en/main_classes/tokenizer).

For our purposes though, since we don't need to support all the models, as long as the tokenizers are available the HF hub i think `tokenizers` library is enough.

File: tests/unit_tests/test_tokenizer.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+            hf_token=None,  # Public model, no token needed
+        )
+
+        # Step 2: Load tokenizer using our function
+        model_name = test_repo_id.split("/")[-1]
+        tokenizer_path = os.path.join(self.temp_dir, model_name)
+        our_tokenizer = load_tokenizer(tokenizer_path)
+
+        # Step 3: Load tokenizer using official Tokenizer library
+        official_tokenizer = Tokenizer.from_pretrained(test_repo_id)
```
Comment (tianyu-l) [inline tests/unit_tests/test_tokenizer.py]: I think CI won't have access to internet.

File: torchtitan/components/tokenizer.py [hunk 5]
```diff
@@ -25,3 +31,333 @@ def decode(self, *args, **kwargs) -> str:
...
+    def token_to_id(self, token: str) -> Optional[int]:
+        """Convert token to ID."""
+        return self.tokenizer.token_to_id(token)
+
+    def id_to_token(self, token_id: int) -> Optional[str]:
+        """Convert ID to token."""
+        return self.tokenizer.id_to_token(token_id)
+
+
+def load_tokenizer_config(tokenizer_path: str) -> dict:
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: is this used?

File: torchtitan/components/tokenizer.py [hunk 6]
```diff
@@ -25,3 +31,333 @@ def decode(self, *args, **kwargs) -> str:
...
+    """Load tokenizer configuration from tokenizer_config.json if it exists."""
+    config_path = os.path.join(tokenizer_path, "tokenizer_config.json")
+    if os.path.exists(config_path):
+        with open(config_path, "r") as f:
+            return json.load(f)
+    return {}
+
+
+def load_tokenizer(tokenizer_path: str) -> TorchTitanTokenizer:
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]: we can mimic the naming for data loader and call it `build_hf_tokenizer`
https://github.com/pytorch/torchtitan/blob/main/torchtitan/datasets/hf_datasets.py#L167

File: torchtitan/components/tokenizer.py [hunk 7]
```diff
@@ -5,7 +5,13 @@
...
+import json
+import os
 from abc import ABC, abstractmethod
+from typing import Any, Optional
+
+from tokenizers import Tokenizer as HfTokenizer
+from typing_extensions import override
 
 
 class Tokenizer(ABC):
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py:17]:
```suggestion
class BaseTokenizer(ABC):
```
similar to https://github.com/pytorch/torchtitan/blob/main/torchtitan/components/dataloader.py#L26

File: torchtitan/components/tokenizer.py [hunk 8]
```diff
@@ -25,3 +31,333 @@ def decode(self, *args, **kwargs) -> str:
     @property
     def n_words(self) -> int:
         return self._n_words
+
+
+class TorchTitanTokenizer(Tokenizer):
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py]:
```suggestion
class Tokenizer(BaseTokenizer):
```
or `HuggingFaceTokenizer` similar to https://github.com/pytorch/torchtitan/blob/main/torchtitan/datasets/hf_datasets.py#L71

**Review by tianyu-l — COMMENTED** [review]
I'm still trying to comprehend the tokenizer space ...

Can we substitute https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/__init__.py#L82 with the new function in this PR and get identical results?

I was wondering if the change is general enough, as I saw torchtune has this folder https://github.com/pytorch/torchtune/tree/main/torchtune/modules/transforms/tokenizers

File: tests/unit_tests/test_tokenizer.py (hunk 4 above)
Comment (H-Huang) [reply tests/unit_tests/test_tokenizer.py]: It looks like the test rand and passed so that means its working? https://github.com/pytorch/torchtitan/actions/runs/15788713163/job/44510547390?pr=1318#step:14:273

File: torchtitan/components/tokenizer.py (hunk 5 above)
Comment (H-Huang) [reply torchtitan/components/tokenizer.py]: oops thanks, removed it

File: torchtitan/components/tokenizer.py (hunk 6 above)
Comment (H-Huang) [reply torchtitan/components/tokenizer.py]: good idea, will update the name

File: torchtitan/components/tokenizer.py (hunk 8 above)
Comment (H-Huang) [reply torchtitan/components/tokenizer.py]: Good idea, I like `HuggingFaceTokenizer` so there is less of a naming conflict with the `Tokenizer` from the `tokenizers` lib

File: torchtitan/components/tokenizer.py (hunk 7 above)
Comment (H-Huang) [reply torchtitan/components/tokenizer.py:17]: Agree on this, I had originally updated this to `BaseTokenizer`, but it also required a lot of changes to many parts of the code that reference this class including a lot of changes in the `experiments/`. So for better understanding of this PR I split these changes off.

I can add it back in if you like, but otherwise would prefer to have it in a separate refactor PR.

**Comment by H-Huang** [comment]
Hi @tianyu-l, some additional context

> I'm still trying to comprehend the tokenizer space ...

Spiritually you can think of this `HuggingFaceTokenizer` implementation as the alternative to `AutoTokenizer` from the `transformers` library. The OSS deepseek model is using AutoTokenizer will download a tokenizer data (`tokenizer.json`, `tokenizer.model` or `vocab.json` or `vocab.txt`, etc.) then apply some [configuration](https://huggingface.co/deepseek-ai/DeepSeek-V3/blob/main/tokenizer_config.json) (for example in DSv3, the tokenizer always adds "beginning of sentence" token to the encoding)

The only difference is in our implementation we split up the steps to 1) `download`, then 2) `build`. But functionally it should have encoding/decodings which are no different from AutoTokenizer.

> Can we substitute https://github.com/pytorch/torchtitan/blob/main/torchtitan/models/llama3/__init__.py#L82 with the new function in this PR and get identical results?

It should, in my unit test i validate that this against DeepSeekv3. But I was denied access to llama3, so im trying to unblock myself there to test

> I was wondering if the change is general enough, as I saw torchtune has this folder https://github.com/pytorch/torchtune/tree/main/torchtune/modules/transforms/tokenizers

I might be missing some things but from other tokenizers I tested it also seemed to work. As long as the models we want to support in torchtitan are well tested I think this should be okay? When loading the tokenizer files there are multiple different strategies which will change what tokenizer is loaded based off the files that are available (https://github.com/pytorch/torchtitan/pull/1318/files#diff-87d941cc62c11fe923ef1118a1b2e8319c9126a775b6151fe539e1472d0656a0R92-R136). If I understand correctly that is basically what torchtune is doing, albeit maybe they cover more edge cases?

**Comment by fegin** [comment]
> It should, in my unit test i validate that this against DeepSeekv3. But I was denied access to llama3, so im trying to unblock myself there to test

I could not access llama3 dataset anymore. I unblock myself by accessing the dataset internally and put it to the expect folder. This works with the existing Tokenizer. I would expect that the new one has the same feature (no need to download if one exists).

File: tests/unit_tests/test_tokenizer.py [hunk 9]
```diff
@@ -0,0 +1,410 @@
...
+                our_tokenizer.tokenizer, official_tokenizer, test_repo_id
+            )
+
+        # Step 6: Compare with transformers tokenizer if available
+        if transformers_tokenizer:
+            self._compare_tokenizers(
+                our_tokenizer, transformers_tokenizer, test_repo_id
+            )
+
+    def test_backward_comptability(self):
```
Comment (H-Huang) [inline tests/unit_tests/test_tokenizer.py:311]: validating that `TikTokenizer` (old implementation) encode and decodes the same as `HuggingFaceTokenizer` (new implementation). Will remove this test after we remove TikTokenizer

File: torchtitan/components/tokenizer.py (hunk 7 above)
Comment (tianyu-l) [reply torchtitan/components/tokenizer.py:17]: Makes sense. Let's do it in a refactor PR.

File: torchtitan/components/tokenizer.py [hunk 10]
```diff
@@ -25,3 +31,402 @@ def decode(self, *args, **kwargs) -> str:
...
+        for key in standard_keys:
+            token_config = self.config.get(key)
+            if token_config is not None:
+                token_str = self._get_token_from_config(self.config, key)
+                if token_str is not None:
+                    added_token = self._process_special_token(token_str, token_config)
+                    added_tokens_to_add.append(added_token)
+
+        # Process added_tokens_decoder (comprehensive special token definitions)
+        added_tokens_decoder = self.config.get("added_tokens_decoder", {})
```
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py:281]: a bit surprised by this: since it's already in the `tokenizer_config.json`, do we still need to call `self.tokenizer.add_special_tokens(added_tokens_to_add)` to add them?
I would assume they can directly put them into `tokenizer.json` or any orginal files...

File: torchtitan/components/tokenizer.py
Comment (tianyu-l) [inline torchtitan/components/tokenizer.py:1]: Not sure what's the most reasonable thing to do, I'd suggest we follow existing file structure for dataloader:
- keep base `tokenizer.py` still in `torchtitan/components/` folder
- put `hf_tokenizer.py` into `torchtitan/datasets` folder.

File: tests/unit_tests/test_tokenizer.py
Comment (tianyu-l) [inline tests/unit_tests/test_tokenizer.py:1]: It looks to me that test in this file will never be run in CI due to no internet access, but I'm not sure where to put it o/w. Maybe we can still put here, so if people run test locally it'd help identify issues.
Let's comment this somewhere

File: tests/unit_tests/test_tokenizer.py [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+        4. Compares with transformers AutoTokenizer (if available)
+        """
+        # Step 1: Download tokenizer files
+        try:
+            download_hf_tokenizer_files(
+                repo_id=test_repo_id,
+                local_dir=self.temp_dir,
+            )
+        except HTTPError as e:
+            if test_repo_id == "meta-llama/Meta-Llama-3.1-8B":
```
Comment (tianyu-l) [inline tests/unit_tests/test_tokenizer.py:270]: what happens if it's deepseek / qwen? I expected those test to raise exceptions but seems not

File: README.md (hunk 3 above)
Comment (tianyu-l) [inline README.md]: for now let's still put it under `torchtitan/models/deepseek_v3/README.md`, as other parts including `Start a training run` in main README need to be updated together.

**Review by tianyu-l — COMMENTED** [review]
Since this PR is not deepseek-specific, I think we can land it in main branch, assuming the deepseek tokenizer works fine in training loop.

We also need to deprecate the old tokenizer approach, could be in a separate PR.
For that, the transition code I added before is at https://github.com/pytorch/torchtitan/blob/main/torchtitan/config_manager.py#L829
We can just update that, substituting both the old and the new.

File: tests/unit_tests/test_tokenizer.py (hunk 11 above)
Comment (H-Huang) [reply tests/unit_tests/test_tokenizer.py:270]: Since these are open-sourced, they don't require a Hugging Face token to be downloaded. I added this condition for CI because Meta's LLaMA model does require a token. However, I can still test this on my machine because the Hugging Face CLI automatically saves your token locally.

File: README.md (hunk 3 above)
Comment (H-Huang) [reply README.md]: makes sense, will update

File: tests/unit_tests/test_tokenizer.py
Comment (H-Huang) [reply tests/unit_tests/test_tokenizer.py:1]: It does run for qwen and deepseek: https://github.com/pytorch/torchtitan/actions/runs/15836819448/job/44642005002?pr=1318#step:14:253

For llama it doesnt because it requires a token

File: torchtitan/components/tokenizer.py (hunk 10 above)
Comment (H-Huang) [reply torchtitan/components/tokenizer.py:281]: Most of the time these special tokens are already included in the `tokenizer.json` file.

My interpretation of the `tokenizer_config.json` is as an overrider. So if you specify these tokens are special then we should make sure that the tokenizer recognizes that.

According to the API (https://huggingface.co/docs/tokenizers/v0.20.3/en/api/tokenizer#tokenizers.Tokenizer.add_special_tokens), "If these tokens are already part of the vocabulary, it just let the Tokenizer know about them. If they don’t exist, the Tokenizer creates them, giving them a new id. " In other words, if they already exist in vocab, then just make sure they are marked as "special" and don't add anything new. Otherwise if they do not exist, increase the vocab and add them. So i think this is pretty safe operation.

The "special" marker is needed because it is later used in `skip_special_tokens` for the decode part (https://huggingface.co/docs/tokenizers/v0.20.3/en/api/tokenizer#tokenizers.Tokenizer.decode.skip_special_tokens)

File: torchtitan/components/tokenizer.py
Comment (H-Huang) [reply torchtitan/components/tokenizer.py:1]: Do you think the `HuggingFaceTokenizer` is general enough to just keep in the `tokenizer.py` file under components? I feel like it is. It only takes a path in its constructor and doesn't really have any model specific things. Users could extend this tokenizer themselves, or use the BaseTokenizer ABC

**Comment by H-Huang** [comment]
> Since this PR is not deepseek-specific, I think we can land it in main branch, assuming the deepseek tokenizer works fine in training loop.

Changed the merge into to the main branch

File: torchtitan/components/tokenizer.py
Comment (tianyu-l) [reply torchtitan/components/tokenizer.py:1]: OK agree, sounds cleaner

File: tests/unit_tests/test_tokenizer.py
Comment (tianyu-l) [reply tests/unit_tests/test_tokenizer.py:1]: oh our CI has internet access? I must have forgotten. Yeah sounds good!

File: torchtitan/components/tokenizer.py (hunk 10 above)
Comment (tianyu-l) [reply torchtitan/components/tokenizer.py:281]: great to know, thanks!

File: scripts/download_tokenizer.py [hunk 12]
```diff
@@ -9,57 +9,161 @@
...
+    # Extract model name from repo_id (part after "/")
+    if "/" not in repo_id:
+        raise ValueError(
+            f"Invalid repo_id format: '{repo_id}'. Expected format: 'organization/model-name'"
         )
+    model_name = repo_id.split("/")[-1].strip()
+    model_dir = os.path.join(local_dir, model_name)
+
+    # Tokenizer file patterns to match (case-insensitive)
+    tokenizer_patterns = [
```
Comment (wwwjn) [inline scripts/download_tokenizer.py:56]: Do we know which of these files are required and which are optional? In some rare cases, the repo is missing some of the files. Torchtitan users will download tokenizer successfully, but can not load the tokenizer

File: tests/unit_tests/test_tokenizer.py [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+
+            self.assertEqual(
+                our_decoded,
+                reference_decoded,
+                f"Decoded text should match for '{text}' in {test_repo_id} ({tokenizer_type})",
+            )
+
+        # 5. Edge case testing
+        edge_cases = [
+            "🚀🌟✨",  # Emojis
```
Comment (wwwjn) [inline tests/unit_tests/test_tokenizer.py:218]: This test is really great!

File: tests/unit_tests/test_tokenizer.py [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+            official_tokenizer = Tokenizer.from_pretrained(test_repo_id)
+        except Exception as e:
+            print(f"Warning: Could not load official tokenizer for {test_repo_id}: {e}")
+
+        # Step 4: Load tokenizer using transformers AutoTokenizer (if available)
+        transformers_tokenizer = None
+        try:
+            from transformers import AutoTokenizer
+
+            transformers_tokenizer = AutoTokenizer.from_pretrained(test_repo_id)
```
Comment (wwwjn) [inline tests/unit_tests/test_tokenizer.py:295]: Will this test still needs to depend on `transformers` package? Do we need to add  `transformers` as a dependency in `requirement-dev.txt` to make this test work in CI?

File: torchtitan/components/tokenizer.py [hunk 15]
```diff
@@ -25,3 +31,395 @@ def decode(self, *args, **kwargs) -> str:
...
+    from various file formats and infers special token behavior.
+
+    Args:
+        tokenizer_path (str): Path to the directory containing tokenizer files.
+                             Should contain one or more of the supported file types.
+
+    Returns:
+        tokenizer (HuggingFaceTokenizer): Loaded tokenizer instance with intelligent BOS/EOS handling
+    """
+    tokenizer = HuggingFaceTokenizer(tokenizer_path)
```
Comment (wwwjn) [inline torchtitan/components/tokenizer.py:424]: Great work! I would change FLUX tokenizer to our tokenizer after this diff landed

File: .ci/docker/requirements.txt [hunk 16]
```diff
@@ -8,3 +8,4 @@ tabulate
 wandb
 fsspec
 tyro
+tokenizers >= 0.15.0
```
Comment (tianyu-l) [inline .ci/docker/requirements.txt:11]: can we remove `tiktoken`?
also please update https://github.com/pytorch/torchtitan/pull/1364

File: scripts/download_tokenizer.py (hunk 12 above)
Comment (tianyu-l) [reply scripts/download_tokenizer.py:56]: I think the idea is -- a tokenizer won't need more than these files, although it may need a subset of them, in which case the download would still be successful

File: tests/unit_tests/test_tokenizer.py [hunk 17]
```diff
@@ -0,0 +1,410 @@
...
+        # Step 2: Load tokenizer using our function
+        model_name = test_repo_id.split("/")[-1]
+        tokenizer_dir = "tokenizer" if model_name == "FLUX.1-dev" else "."
+        tokenizer_path = os.path.join(self.temp_dir, model_name, tokenizer_dir)
+        our_tokenizer = build_hf_tokenizer(tokenizer_path)
+
+        # Step 3: Load tokenizer using official Tokenizer library (if available)
+        official_tokenizer = None
+        try:
+            official_tokenizer = Tokenizer.from_pretrained(test_repo_id)
```
Comment (tianyu-l) [inline tests/unit_tests/test_tokenizer.py:286]: might be dumb question -- could you remind me of the reasons why we don't just use this method but write our own?

I understand for HF `transformers` it's too big of a dependency, but we anyways need to depend on HF `tokenizers`.

**Review by tianyu-l — APPROVED** [review]
lgtm -- looks very solid!

please address final comments before merge

File: scripts/download_tokenizer.py (hunk 12 above)
Comment (H-Huang) [reply scripts/download_tokenizer.py:56]: Correct, these are just the files that the download script looks for, but we only need a subset to actually initialize the tokenizer. Typically we only need `tokenizer.json` for initialize the tokenizer.

File: tests/unit_tests/test_tokenizer.py (hunk 14 above)
Comment (H-Huang) [reply tests/unit_tests/test_tokenizer.py:295]: That's a good idea, I tested it locally since I have it installed but I think it makes sense to have it in dev as well.

File: .ci/docker/requirements.txt (hunk 16 above)
Comment (H-Huang) [reply .ci/docker/requirements.txt:11]: Will remove it in the follow up PR https://github.com/pytorch/torchtitan/pull/1333

File: tests/unit_tests/test_tokenizer.py (hunk 17 above)
Comment (H-Huang) [reply tests/unit_tests/test_tokenizer.py:286]: This will download the `tokenizer.json` and use it, but it does not apply `tokenizer_config.json` nor does it handle other types of configurations (e.g. `vocab.json`). Our implementation takes these extra cases into account.
